File: Bronze/2018/January/02_lifeguards/Lifeguards3.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Red
 * @version 1.0
 */
public class Lifeguards3 {
    static List<List<Integer>> liveGuards;
    static int[] times = new int[1001];
    static int covered = 0;

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("lifeguards.in"));
        PrintWriter pw = new PrintWriter("lifeguards.out");
        int n = Integer.parseInt(r.readLine());
        liveGuards = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            String[] s = r.readLine().split(" ");
            List<Integer> liveGuard = new ArrayList<>();
            liveGuard.add(Integer.parseInt(s[0]));
            liveGuard.add(Integer.parseInt(s[1]));
            liveGuards.add(liveGuard);
            for (int t = liveGuard.get(0); t < liveGuard.get(1); t++) {
                if (times[t] == 0) {
                    covered++;
                }
                times[t]++;
            }
        }
        int ans = 0;
        // 依次计算开除救生员 i 之后可以覆盖的时间
        for (int i = 0; i < liveGuards.size(); i++) {
            ans = Math.max(ans, coverTimeMissOne(i));
            if (ans == covered) {
                break;
            }
        }
        pw.println(ans);
        pw.close();
    }

    /**
     * 计算开除指定救生员后可以覆盖的时间
     */
    static int coverTimeMissOne(int missed) {
        int[] timesCopy = times.clone();
        int coveredCopy = covered;
        for (int i = liveGuards.get(missed).get(0); i < liveGuards.get(missed).get(1); i++) {
            if (timesCopy[i] == 1) {
                coveredCopy--;
            }
            timesCopy[i]--;
        }
        return coveredCopy;
    }
}

File: Bronze/2018/January/02_lifeguards/lifeguards3.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> live_guards;
int times[1001];
int covered = 0;
// 计算开除指定救生员后可以覆盖的时间
int cover_time_miss_one(int missed) {
    int times_copy[1001];
    copy(begin(times), end(times), begin(times_copy));
    int covered_copy = covered;
    for (int i = live_guards[missed][0]; i < live_guards[missed][1]; i++) {
        if (times_copy[i] == 1) {
            covered_copy--;
        }
        times_copy[i]--;
    }
    return covered_copy;
}

int main() {
    ifstream fin("lifeguards.in");
    ofstream fout("lifeguards.out");

    int n;
    fin >> n;

    live_guards.resize(n);
    for (int i = 0; i < n; i++) {
        live_guards[i].resize(2);
        fin >> live_guards[i][0] >> live_guards[i][1];
        // 依次计算开除救生员 i 之后可以覆盖的时间
        for (int t = live_guards[i][0]; t < live_guards[i][1]; t++) {
            if (times[t] == 0) {
                covered++;
            }
            times[t]++;
        }
    }

    int ans = 0;

    for (int i = 0; i < live_guards.size(); i++) {
        ans = max(ans, cover_time_miss_one(i));
        if (ans == covered) {
            break;
        }
    }

    fout << ans << endl;

    fin.close();
    fout.close();

    return 0;
}
